### ISSUE TITLE
Event Hubs / Executor Performance Investigation

### ISSUE THREAD

**Issue by mamaso** [Azure/azure-webjobs-sdk#1060]
Currently, the function executor creates a large number of timers with relatively long durations.

This isn't a problem for long running functions, however in the case of many short-lived functions (IoT, event hub, service bus) the timers are not doing meaningful work and are causing lock contention (in the .net TimerQueue implementation), as well as possibly when callbacks fire and post back to the threadpool (less convinced that this is an issue).

A couple improvements that could be made to the executor:

- maintain a measure of function execution time (perhaps moving average). If the execution time is low enough, remove certain timers (update log, update parameter log, etc)
- add timer jitter to make these timer callbacks fall in different batches

There could also be:
- improve clarity, it's currently very hard to follow the logic
- move any initialization logic out of executor into indexer

Implementation of .net timers can be found here: https://referencesource.microsoft.com/#mscorlib/system/threading/timer.cs,b15414c13eba8b8f

**Comment by mamaso** [comment]
Updates:

I haven't found the expected contention in the timer pool - that doesn't appear to be an issue.

After doing some performance profiling, I found that we spend ~25% of execution in reflection calls. I moved those to the indexing path in #1102 so that they only run at initialization.

I found a few improvements related to contention after profiling as well:
- Storage ops are a significant source of contention. They create a number of worker threads which contend in a [custom finalizer](https://github.com/Azure/azure-storage-net/blob/afc5c6d99a805fb4165631ab5cfd139238b00e71/Lib/ClassLibraryCommon/Core/Util/AsyncStreamCopier.cs#L165). These operations primarily come from:
  - FastTable logging
    - When removing the dashboard connection string to disable logging we can see throughput improvements of ~25%. For high throughput scenarios like event hubs table logging has a high operation latency (~500ms) and high volume (GBs per hour).
  - EventProcessorHost
    - Event processor host does a lot of storage operations under the hood for partition lease managment and checkpointing. We have control over checkpointing - decreasing the checkpoint frequency increased throughput 10-15%.
- Storage ops in dashboard logging path should be async (see [here](https://github.com/Azure/azure-webjobs-sdk/blob/663a508e8a851629c26a51e7de3af36629dfd120/src/Microsoft.Azure.WebJobs.Logging/Entities/ContainerActiveEntity.cs#L78))
